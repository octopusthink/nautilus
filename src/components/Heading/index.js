import { css } from '@emotion/core';
import invariant from 'invariant';
import PropTypes from 'prop-types';
import React from 'react';

import { useTheme } from 'themes';
import { headingLarge, headingMedium, headingSmall } from 'themes/mixins';

export const Heading = ({
  children,
  level,
  large,
  medium,
  small,
  ...otherProps
}) => {
  invariant(level, 'Heading component requires a level prop');

  const HeadingElement = `h${level}`;
  const theme = useTheme();

  return (
    <HeadingElement
      css={css`
        ${css(headingMedium(theme))};
        margin: 0 0 ${theme.spacing.margin.m};

        ${large &&
          css`
            ${css(headingLarge(theme))}
          `}

        ${small &&
          css`
            ${css(headingSmall(theme))}
          `}
      `}
      {...otherProps}
    >
      {children}
    </HeadingElement>
  );
};

export const HeadingLevels = [1, 2, 3, 4, 5, 6];

Heading.defaultProps = {
  children: undefined,
  level: 2,
  large: true,
  medium: false,
  small: false,
};

Heading.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Semantic hierarchy level of the `<h>` element in the markup (ex: `<h3>`) */
  level: PropTypes.oneOf(HeadingLevels).isRequired,
  /** Visual prominency of the heading. Small, medium, large. */
  large: PropTypes.bool,
  medium: PropTypes.bool,
  small: PropTypes.boolm,
};

export default Heading;
