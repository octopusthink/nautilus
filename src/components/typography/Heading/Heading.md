## Available Sizes
### XXLarge Heading
```js
<Heading level={2} size="xxlarge">This is a heading</Heading>
```

### XLarge Heading
```js
<Heading level={2} size="xlarge">This is a heading</Heading>
```

### Large Heading
```js
<Heading level={2} size="large">This is a heading</Heading>
```

### Medium Heading
```js
<Heading level={2}>This is a heading</Heading>
```

### Small Heading
```js
<Heading level={2} size="small">This is a heading</Heading>
```

### XSmall Heading
```js
<Heading level={2} size="xsmall">This is a heading</Heading>
```

### XXSmall Heading
```js
<Heading level={2} size="xxsmall">This is a heading</Heading>
```

## Colors
### Headings with custom colors
```js
<Heading level={2} color="light">This is a heading</Heading>
<Heading level={2} color="medium">This is a heading</Heading>
<Heading level={2} color="dark">This is a heading</Heading>
<Heading level={2} color="accent">This is a heading</Heading>
<Heading level={2} color="danger">This is a heading</Heading>
<Heading level={2} color="success">This is a heading</Heading>
```
