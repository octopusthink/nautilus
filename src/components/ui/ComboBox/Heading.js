import { css } from '@emotion/core';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import { toUnits } from '../../../styles';
import { useTheme } from '../../../themes';
import ListItem from '../List/Item';

export const ComponentClassName = 'Nautilus-ComboBoxHeading';

export const Heading = (props) => {
  const { children, className, unstyled, ...otherProps } = props;

  const theme = useTheme();

  return (
    <ListItem
      className={classnames(ComponentClassName, className)}
      css={
        unstyled
          ? undefined
          : css`
              background: ${theme.colors.text.default};
              color: ${theme.colors.text.inverse};
              line-height: 2.4rem;
              padding: ${toUnits(theme.spacing.padding.medium)};
              width: 100%;
            `
      }
      unstyled
      {...otherProps}
    >
      {children}
    </ListItem>
  );
};

Heading.defaultProps = {
  children: undefined,
  className: undefined,
  unstyled: false,
};

Heading.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  className: PropTypes.string,
  /* @ignore Don't output any CSS styles. */
  unstyled: PropTypes.bool,
};

export const { defaultProps, propTypes } = Heading;

export default Heading;
