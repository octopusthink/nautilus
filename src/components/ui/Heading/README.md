Headings are short pieces of text used to organise and structure a page's content.

They provide both a semantic and a presentational role, guiding users (as well as search engines and other bots) through the page content both visually (through hierarchy) and structurally (through markup). They are especially important for users of assistive technology—heading allow quick navigation throughout a page.

```jsx
<Heading>This just in: Scotland is rainy 🏴󠁧󠁢󠁳󠁣󠁴󠁿🌦</Heading>
```

## Usage

Use a `<Heading>` to introduce a block of text or other content. Headings act as titles for sections of content, and are almost always followed by more detailed information.

Don't use a `<Heading>` for:
- The primary title of a page. Use a `<PageTitle>` instead.
- A subtitle underneath a primary title. Use a `<Subtitle>` instead.
- Long blocks of text intended to span more than two lines. Use a [`<Paragraph>`](../#/Function/Components/Paragraph) instead.
- A primary link that stands by itself. Use a [`<Button>`](../#/Function/Components/Button) instead.

## Appearance

- Use a consistent colour and style for headings across the site.
- Use variations in size to establish hierarchy in headings.
- The space between a heading and the content it titles should be notably smaller than the space between the heading and its preceding content. This helps establish a relationship by ensuring that headings are in close proximity to the content they describe.
- Headings shouldn't span longer than two lines.
- Headings should be visibly larger and more prominent than body copy, and should be more tightly spaced between lines.

## Variations

Headings are provided in three different sizes. You should **always use headings sequentially** (a medium heading would be nested underneath a large heading).

```jsx
<Heading level={2}>Breaking news 📰</Heading>
<Heading level={3}>Kittens are better than dogs 😼</Heading>
<Heading level={4}>More at eleven 🕚</Heading>
```

Headings are also available in dark and light variants, as well as an inverse. These colours can be set in your theme.

```jsx
<Heading level={4} dark>Dark</Heading>
<Heading level={4}>Default</Heading>
<Heading level={4} light>Light</Heading>
<Heading level={4} inverse dark>Inverse Dark</Heading>
<Heading level={4} inverse>Inverse</Heading>
<Heading level={4} inverse light>Inverse Light</Heading>
```

## Voice & Tone

- Headings should use Sentence case, not Title Case.
- Headings should use a full stop (period) only when they are written as full sentences.
- Keep headings as short as possible. Only use a full sentence for a heading when necessary for stylistic purposes.
- Use headings to separate and organise long blocks of content.
- Headings should reflect and title the content that follows them.

## Accessibility

For accessibility reasons, it's important that `<Heading>` components follow proper semantic order.

- Never skip a heading level (ie, going from `<h2>` to `<h4>`).
- Don't traverse heading levels (ie, nesting an `<h2>` under an `<h4>`). Headings should be used sequentially to provide further separation inside an existing section.
- Only use a single `<h1>` tag per page. (This should be your `<PageTitle>`.)
- Don't write headings in ALL CAPS. The variation of font, size, and/or weight should be a sufficient indicator of hierarchy.
