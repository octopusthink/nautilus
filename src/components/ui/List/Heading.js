import React from 'react';

export const ListHeading = (props) => {
  return <React.Fragment {...props} />;
};

export default ListHeading;
