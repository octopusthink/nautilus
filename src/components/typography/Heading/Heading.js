import PropTypes from 'prop-types';
import React from 'react';

import { headingSizes, headingColors, headingElements } from './Heading.styles';

export const Heading = ({ children, color, level, size, ...others }) => {
  const HeadingElement = headingElements[`h${level}`];
  return (
    <HeadingElement
      color={color}
      size={size}
      {...others}
    >
      {children}
    </HeadingElement>
  );
};

export const HeadingColors = Object.keys(headingColors);
export const HeadingLevels = [1, 2, 3, 4, 5, 6];
export const HeadingSizes = Object.keys(headingSizes);

Heading.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Color of the Heading. Options are 'dark', 'medium', 'light', 'accent', 'danger', and 'success' */
  color: PropTypes.oneOf(HeadingColors),
  /** Hierarchy level of the Heading (ex: h3) */
  level: PropTypes.oneOf(HeadingLevels),
  /** Size of the Heading. Options are 'xxsmall', 'xsmall', 'small', 'medium', 'large', 'xlarge', 'xxlarge' */
  size: PropTypes.oneOf(HeadingSizes),
};

Heading.defaultProps = {
  color: 'dark',
  level: 2,
  size: 'medium',
};
