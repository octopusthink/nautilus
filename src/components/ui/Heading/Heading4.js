import styled from '@emotion/styled';
import React from 'react';

import Heading from '.';

// We define `<Heading4>` for developer convenience.
export const Heading4 = (props) => {
  return <Heading level={4} {...props} />;
};

export default styled(Heading4)``;
