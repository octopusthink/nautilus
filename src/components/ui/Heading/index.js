import { css } from '@emotion/core';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import { ComponentClassName as ListClassName } from 'components/ui/List';
import { ComponentClassName as ParagraphClassName } from 'components/ui/Paragraph';
import { heading, toUnits } from 'styles';
import { useTheme } from 'themes';

const LARGE = 2;
const MEDIUM = 3;
const SMALL = 4;

const ComponentClassName = 'Nautilus-Heading';
const HeadingLevels = [LARGE, MEDIUM, SMALL];

export const headingClassNameForLevel = (level) => {
  return `${ComponentClassName}-${level}`;
};

export const Heading = (props) => {
  const theme = useTheme();
  const { children, className, level, ...otherProps } = props;
  const HeadingElement = `h${level}`;

  return (
    <HeadingElement
      className={classnames(headingClassNameForLevel(level), className)}
      css={css`
        margin: 0 0 ${toUnits(theme.spacing.margin.medium)};
        .${ListClassName} + &,
        .${ParagraphClassName} + &,
        /* TODO: Replace these with actual Nautilus components */
        dl + & {
          margin-top: ${toUnits(theme.spacing.margin.medium)};
        }

        ${level === SMALL && heading.small(theme)};
        ${level === MEDIUM && heading.medium(theme)};
        ${level === LARGE && heading.large(theme)};
      `}
      {...otherProps}
    >
      {children}
    </HeadingElement>
  );
};

Heading.defaultProps = {
  className: undefined,
  children: undefined,
  level: 2,
};

Heading.propTypes = {
  /** @ignore */
  className: PropTypes.string,
  /** @ignore */
  children: PropTypes.node,
  /** Semantic hierarchy level of the `<h>` element in the markup (ex: `<h3>`). The more semantically important the level, the larger the heading will appear visually; an `<h2>` will be visually styled as "large" while an `<h4>` will be visually small. */
  level: PropTypes.oneOf(HeadingLevels),
};

export default Heading;
