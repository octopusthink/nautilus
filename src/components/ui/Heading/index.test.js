import React from 'react';

import { axe, muteConsole, render } from '../../../../utils/testing';
import Heading from '.';

describe('Heading', () => {
  it('should render an <h2> by default', () => {
    const { container } = render(<Heading>Hello</Heading>);

    expect(container.firstChild.tagName).toEqual('H2');
  });

  it('should output its children', () => {
    const { getByTestId } = render(
      <Heading>
        <div data-testid="child" />
      </Heading>,
    );

    expect(getByTestId('child')).toBeDefined();
  });
  it('should render an <h2> tag when the `level` prop is 2', () => {
    const { container } = render(<Heading level={2}>Big heading</Heading>);

    expect(container.firstChild.tagName).toEqual('H2');
  });

  it('should render an <h3> tag when the `level` prop is 3', () => {
    const { container } = render(<Heading level={3}>So-so heading</Heading>);

    expect(container.firstChild.tagName).toEqual('H3');
  });

  it('should render an <h4> tag when the `level` prop is 4', () => {
    const { container } = render(<Heading level={4}>Tiny heading</Heading>);

    expect(container.firstChild.tagName).toEqual('H4');
  });

  it('should output large CSS when the `level` prop is 2', () => {
    const { container } = render(<Heading level={2}>Big heading</Heading>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should output medium CSS when the `level` prop is 3', () => {
    const { container } = render(<Heading level={3}>So-so heading</Heading>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should output small CSS when the `level` prop is 4', () => {
    const { container } = render(<Heading level={4}>Tiny heading</Heading>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should invert CSS colour when the `inverse` prop is set', () => {
    const { container } = render(<Heading inverse>Inverted heading</Heading>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should darken text colour when `dark` prop is set', () => {
    const { container } = render(<Heading dark>Inverted heading</Heading>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should only allow one of dark/light props', () => {
    expect(() => {
      muteConsole({ times: 2, type: 'error' });
      render(
        <Heading dark light>
          Black and white
        </Heading>,
      );
    }).toThrow('Heading cannot have both `dark` and `light` props set.');
  });

  it('should only allow one of large/small props', () => {
    expect(() => {
      muteConsole({ times: 2, type: 'error' });
      render(
        <Heading large small>
          Big and little
        </Heading>,
      );
    }).toThrow('Heading cannot have both `large` and `small` props set.');
  });

  it('should accept and pass through other props', () => {
    const { getByTestId } = render(
      <Heading className="custom-class" data-testid="heading" level={3}>
        hello
      </Heading>,
    );

    expect(getByTestId('heading').classList).toContain('custom-class');
  });

  it('should not output styles when unstyled is set', () => {
    const { container } = render(<Heading unstyled>hello</Heading>);

    expect(container).toMatchSnapshot();
  });

  it('should not output styles when unstyled is set, even when style props are set', () => {
    const { container } = render(
      <Heading level={4} unstyled>
        hello
      </Heading>,
    );

    expect(container).toMatchSnapshot();
  });

  it("shouldn't have a margin when `noMargin` is true", () => {
    const { container } = render(<Heading noMargin>Hello</Heading>);

    expect(container.firstChild).toMatchSnapshot();
  });

  describe('accessibility', () => {
    it('should pass aXe tests', async () => {
      const { container } = render(<Heading level={3}>So-so heading</Heading>);

      expect(await axe(container.innerHTML)).toHaveNoViolations();
    });
  });
});
