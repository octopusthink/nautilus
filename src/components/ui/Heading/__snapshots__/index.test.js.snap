// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading should darken text colour when \`dark\` prop is set 1`] = `
.emotion-0 {
  color: #3b3f45;
  margin: 0;
  margin: 0 0 3.2rem;
  color: #181b1c;
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 600;
  font-size: 3.6rem;
  line-height: 1.2222222222222223;
}

dl + .emotion-0,
.Nautilus-List + .emotion-0,
.Nautilus-Paragraph + .emotion-0 {
  margin-top: 3.2rem;
}

<h2
  class="emotion-0"
>
  Inverted heading
</h2>
`;

exports[`Heading should invert CSS colour when the \`inverse\` prop is set 1`] = `
.emotion-0 {
  color: #3b3f45;
  margin: 0;
  margin: 0 0 3.2rem;
  color: #ebedf0;
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 600;
  font-size: 3.6rem;
  line-height: 1.2222222222222223;
}

dl + .emotion-0,
.Nautilus-List + .emotion-0,
.Nautilus-Paragraph + .emotion-0 {
  margin-top: 3.2rem;
}

<h2
  class="emotion-0"
>
  Inverted heading
</h2>
`;

exports[`Heading should not output styles when unstyled is set 1`] = `
<div>
  <h2
    class="emotion-0"
  >
    hello
  </h2>
</div>
`;

exports[`Heading should not output styles when unstyled is set, even when style props are set 1`] = `
<div>
  <h4
    class="emotion-0"
  >
    hello
  </h4>
</div>
`;

exports[`Heading should output large CSS when the \`level\` prop is 2 1`] = `
.emotion-0 {
  color: #3b3f45;
  margin: 0;
  margin: 0 0 3.2rem;
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 600;
  font-size: 3.6rem;
  line-height: 1.2222222222222223;
}

dl + .emotion-0,
.Nautilus-List + .emotion-0,
.Nautilus-Paragraph + .emotion-0 {
  margin-top: 3.2rem;
}

<h2
  class="emotion-0"
>
  Big heading
</h2>
`;

exports[`Heading should output medium CSS when the \`level\` prop is 3 1`] = `
.emotion-0 {
  color: #3b3f45;
  margin: 0;
  margin: 0 0 3.2rem;
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 600;
  font-size: 3.1rem;
  line-height: 1.2903225806451613;
}

dl + .emotion-0,
.Nautilus-List + .emotion-0,
.Nautilus-Paragraph + .emotion-0 {
  margin-top: 3.2rem;
}

<h3
  class="emotion-0"
>
  So-so heading
</h3>
`;

exports[`Heading should output small CSS when the \`level\` prop is 4 1`] = `
.emotion-0 {
  color: #3b3f45;
  margin: 0;
  margin: 0 0 3.2rem;
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 600;
  font-size: 2.7rem;
  line-height: 1.3333333333333333;
}

dl + .emotion-0,
.Nautilus-List + .emotion-0,
.Nautilus-Paragraph + .emotion-0 {
  margin-top: 3.2rem;
}

<h4
  class="emotion-0"
>
  Tiny heading
</h4>
`;
